src/rhsmlib/cloud/README.md

```diff
@@ -1,8 +1,189 @@
 RHSM & Cloud
 ============
 
-This package contains modules for detecting cloud providers and collecting cloud metadata. These
-metadata then can be for example reported in system facts. Currently three main cloud providers
-are supported: Amazon Web Services, Microsoft Azure and Google Cloud Platform. If you want to add
-support for another cloud provider, then add subclasses of CloudCollector and CloudDetector to
-some module in providers sub-package and modify list of supported classes in utils.py 
\ No newline at end of file
+This package contains modules for detecting cloud providers and collecting cloud metadata. The
+metadata then can be for example reported in system facts. Three main cloud providers are
+supported ATM: Amazon Web Services, Microsoft Azure and Google Cloud Platform. If you want to add
+support for another cloud provider, then add subclasses of CloudProvider to module in `providers`
+sub-package and modify list of supported classes in `provider.py`.
+
+Example: you want to add support for Foo Cloud Provider. You will create package `fcp.py` in
+folder `providers`. Content of `fcp.py` will look like this:
+
+```python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2021 Foo Company
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+
+"""
+This is module implementing detector and metadata collector of virtual machine running on Foo Cloud Provider
+"""
+
+import logging
+
+from typing import Union
+
+from rhsmlib.cloud._base_provider import BaseCloudProvider
+
+
+log = logging.getLogger(__name__)
+
+
+class FooCloudProvider(BaseCloudProvider):
+    """
+    Base class for Foo cloud provider
+    """
+
+    CLOUD_PROVIDER_ID = "foo"
+
+    CLOUD_PROVIDER_METADATA_URL = "http://1.2.2.4/metadata/"
+
+    CLOUD_PROVIDER_METADATA_TYPE = "application/json"
+
+    CLOUD_PROVIDER_SIGNATURE_URL = "http://169.254.169.254/signature"
+
+    CLOUD_PROVIDER_SIGNATURE_TYPE = "application/json"
+
+    METADATA_CACHE_FILE = None
+
+    SIGNATURE_CACHE_FILE = None
+
+    HTTP_HEADERS = {
+        'user-agent': 'RHSM/1.0',
+    }
+
+    def __init__(self, hw_info):
+        """
+        Initialize instance of FooCloudDetector
+        """
+        super(FooCloudProvider, self).__init__(hw_info)
+
+    def is_vm(self):
+        """
+        Is system running on virtual machine or not
+        :return: True, when machine is running on VM; otherwise return False
+        """
+        return super(FooCloudProvider, self).is_vm()
```

Comment:
this is extra now

Reply:
You never know. Some provider can have something extra.

Reply:
Reimplementing it is IMHO a corner case, not worth being generally documented outside the code docstrings.